relays/bin-substrate/src/cli/derive_account.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::cli::AccountId;
+use relay_substrate_client::Chain;
+use structopt::{clap::arg_enum, StructOpt};
+
+/// Given a source chain `AccountId`, derive the corresponding `AccountId` for the target chain.
+///
+/// The (derived) target chain `AccountId` is going to be used as dispatch origin of the call
+/// that has been sent over the bridge.
+/// This account can also be used to receive target-chain funds (or other form of ownership),
+/// since messages sent over the bridge will be able to spend these.
+#[derive(StructOpt)]
+pub struct DeriveAccount {
+	/// A bridge instance to initalize.
+	#[structopt(possible_values = &DeriveAccountBridge::variants(), case_insensitive = true)]
+	bridge: DeriveAccountBridge,
+	/// Source-chain address to derive Target-chain address from.
+	account: AccountId,
+}
+
+arg_enum! {
+	#[derive(Debug)]
+	/// Bridge to derive account for.
+	pub enum DeriveAccountBridge {
+		MillauToRialto,
+		RialtoToMillau,
+	}
+}
+
+macro_rules! select_bridge {
+	($bridge: expr, $generic: tt) => {
+		match $bridge {
+			DeriveAccountBridge::MillauToRialto => {
+				type Source = relay_millau_client::Millau;
+				type Target = relay_rialto_client::Rialto;
+
+				$generic
+			}
+			DeriveAccountBridge::RialtoToMillau => {
+				type Source = relay_rialto_client::Rialto;
+				type Target = relay_millau_client::Millau;
+
+				$generic
+			}
+		}
+	};
+}
+
+impl DeriveAccount {
+	/// Parse CLI arguments and derive account.
+	///
+	/// Returns both the Source account in correct SS58 format and the derived account.
+	fn derive_account(&self) -> (AccountId, AccountId) {
+		select_bridge!(self.bridge, {
+			let mut account = self.account.clone();
+			account.enforce_chain::<Source>();
+			let acc = bp_runtime::SourceAccount::Account(account.raw_id());
+			let id = bp_millau::derive_account_from_rialto_id(acc);
+			let derived_account = AccountId::from_raw::<Target>(id);
```

Comment:
Should't this be part of the macro code? For example, in the case of Millau to Rialto we would want:

```rust
let id = bp_rialto::derive_account_from_millau_id(acc);
```

Reply:
Great catch, I trusted myself too much that after #849 the code sections are properly split. It didn't trigger any compilation errors, cause the account types are the same on both chains.